### ISSUE TITLE
zipformer2 转成ONNX后，ONNX推理结果有问题

### ISSUE THREAD

**Issue by 0120wzw-coder** [k2-fsa/icefall#2032]
1、训练模型时的流式参数设置：
    --causal true \
    --chunk-size 32 \
    --left-context-frames 128
2、导出流式onnx流式参数为（export-onnx-streaming.py）：
  --causal true \
  --chunk-size 32 \
  --left-context-frames 128

3、使用onnx_pretrained_streaming解码时，音频的转录结果有问题：
`INFO [onnx_pretrained_streaming.py:540] ⁇  ⁇  ⁇  ⁇  ⁇  ⁇ ho ⁇ hola ⁇  ⁇  ⁇  ⁇  ⁇  ⁇ a ⁇  ⁇  ⁇  ⁇  ⁇  ⁇  ⁇  ⁇ bien ⁇ bien ⁇ bien ⁇  ⁇ bienvenido ⁇  ⁇  ⁇  ⁇  ⁇ tiene ⁇ tiene ⁇  ⁇ que ⁇  ⁇ que ⁇  ⁇ haber ⁇  ⁇  ⁇ algo ⁇ algo ⁇  ⁇  ⁇ más ⁇  ⁇ más ⁇  ⁇  ⁇  ⁇  ⁇ gracias ⁇ gracias ⁇  ⁇  ⁇  ⁇  ⁇  ⁇  ⁇ por ⁇ por ⁇  ⁇  ⁇ ven ⁇  ⁇  ⁇  ⁇  ⁇  ⁇ hoy ⁇  ⁇ hoy ⁇  ⁇ a ⁇  ⁇ con ⁇  ⁇  ⁇  ⁇  ⁇  ⁇  ⁇  ⁇  ⁇`

请问这种问题怎么解决呢

**Comment by csukuangfj** [comment]
信息非常有限，没法提供帮助。

建议贴训练和解码最开始的100行log

**Comment by 0120wzw-coder** [comment]
解码的脚本：egs/librispeech/ASR/zipformer/onnx_pretrained-streaming.py
解码的log：
`Sound File: /audio_split/10017/10017_3.wav
Encoder Model: /asr-online_10000/encoder-epoch-30-avg-5-chunk-32-left-128.onnx
Decoder Model: /asr-online_10000/decoder-epoch-30-avg-5-chunk-32-left-128.onnx
Joiner Model: /asr-online_10000/joiner-epoch-30-avg-5-chunk-32-left-128.onnx
Bpe Path: /mls_spanish_opus/gen/spm_es.model
2025-10-15 10:36:31,895 INFO [onnx_pretrained_streaming.py:155] encoder_meta={'num_heads': '4,4,4,8,4,4', 'cnn_module_kernels': '31,31,15,15,15,31', 'encoder_dims': '192,256,384,512,384,256', 'query_head_dims': '32,32,32,32,32,32', 'T': '77', 'value_head_dims': '12,12,12,12,12,12', 'decode_chunk_len': '64', 'left_context_len': '128,64,32,16,32,64', 'num_encoder_layers': '2,2,3,4,3,2', 'comment': 'streaming zipformer2', 'version': '1', 'model_author': 'k2-fsa', 'model_type': 'zipformer2'}
2025-10-15 10:36:31,896 INFO [onnx_pretrained_streaming.py:182] decode_chunk_len: 64
2025-10-15 10:36:31,896 INFO [onnx_pretrained_streaming.py:183] T: 77
2025-10-15 10:36:31,896 INFO [onnx_pretrained_streaming.py:184] num_encoder_layers: [2, 2, 3, 4, 3, 2]
2025-10-15 10:36:31,896 INFO [onnx_pretrained_streaming.py:185] encoder_dims: [192, 256, 384, 512, 384, 256]
2025-10-15 10:36:31,896 INFO [onnx_pretrained_streaming.py:186] cnn_module_kernels: [31, 31, 15, 15, 15, 31]
2025-10-15 10:36:31,896 INFO [onnx_pretrained_streaming.py:187] left_context_len: [128, 64, 32, 16, 32, 64]
2025-10-15 10:36:31,896 INFO [onnx_pretrained_streaming.py:188] query_head_dims: [32, 32, 32, 32, 32, 32]
2025-10-15 10:36:31,896 INFO [onnx_pretrained_streaming.py:189] value_head_dims: [12, 12, 12, 12, 12, 12]
2025-10-15 10:36:31,896 INFO [onnx_pretrained_streaming.py:190] num_heads: [4, 4, 4, 8, 4, 4]
2025-10-15 10:36:31,938 INFO [onnx_pretrained_streaming.py:247] context_size: 2
2025-10-15 10:36:31,938 INFO [onnx_pretrained_streaming.py:248] vocab_size: 4096
2025-10-15 10:36:31,981 INFO [onnx_pretrained_streaming.py:260] joiner_dim: 512
2025-10-15 10:36:31,981 INFO [onnx_pretrained_streaming.py:482] Constructing Fbank computer
2025-10-15 10:36:31,985 INFO [onnx_pretrained_streaming.py:485] Reading sound files: /home/ma-user/work/datasets/ES_AR_datasets/es_A24com/es_A24com_vad/audio_split/10017/10017_3.wav
[1, 1, 1, 1, 7, 1, 7, 1, 1, 1, 14, 1, 14, 1, 1, 1, 9, 1, 1, 1, 3969, 1, 1, 1, 1, 1, 1, 1, 1778, 1, 1, 1, 1778, 68, 1, 1, 1, 1, 5, 1, 5, 1, 1, 1, 1, 1, 1, 1, 490, 1, 1, 5, 1, 5, 1, 1, 1, 101, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3810, 1, 1, 3810, 1, 1, 1, 1, 1, 1, 1]
2025-10-15 10:36:32,629 INFO [onnx_pretrained_streaming.py:539] /home/ma-user/work/datasets/ES_AR_datasets/es_A24com/es_A24com_vad/audio_split/10017/10017_3.wav
2025-10-15 10:36:32,629 INFO [onnx_pretrained_streaming.py:540] ⁇  ⁇  ⁇  ⁇ la ⁇ la ⁇  ⁇  ⁇ del ⁇ del ⁇  ⁇  ⁇ el ⁇  ⁇  ⁇ viernes ⁇  ⁇  ⁇  ⁇  ⁇  ⁇  ⁇ tre ⁇  ⁇  ⁇ trece ⁇  ⁇  ⁇  ⁇ de ⁇ de ⁇  ⁇  ⁇  ⁇  ⁇  ⁇  ⁇ julio ⁇  ⁇ de ⁇ de ⁇  ⁇  ⁇ dos ⁇  ⁇  ⁇  ⁇  ⁇  ⁇  ⁇  ⁇  ⁇  ⁇  ⁇ veinti ⁇  ⁇ veinti ⁇  ⁇  ⁇  ⁇  ⁇  ⁇  ⁇
2025-10-15 10:36:32,629 INFO [onnx_pretrained_streaming.py:542] Decoding Done`

这个：
`[1, 1, 1, 1, 7, 1, 7, 1, 1, 1, 14, 1, 14, 1, 1, 1, 9, 1, 1, 1, 3969, 1, 1, 1, 1, 1, 1, 1, 1778, 1, 1, 1, 1778, 68, 1, 1, 1, 1, 5, 1, 5, 1, 1, 1, 1, 1, 1, 1, 490, 1, 1, 5, 1, 5, 1, 1, 1, 101, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3810, 1, 1, 3810, 1, 1, 1, 1, 1, 1, 1]` 
是预测的token, “1” 在词表中是 “unk”
真实标签为： la del viernes trece de julio de dos mil veinticinco

**Comment by 0120wzw-coder** [comment]
训练日志：
2025-10-15 10:32:36,121 INFO [train.py:1278] (1/8) {
  "am_scale": 0.0,
  "attention_decoder_attention_dim": 512,
  "attention_decoder_dim": 512,
  "attention_decoder_feedforward_dim": 2048,
  "attention_decoder_loss_scale": 0.8,
  "attention_decoder_num_heads": 8,
  "attention_decoder_num_layers": 6,
  "average_period": 200,
  "base_lr": 0.045,
  "batch_idx_train": 0,
  "best_train_epoch": -1,
  "best_train_loss": Infinity,
  "best_valid_epoch": -1,
  "best_valid_loss": Infinity,
  "blank_id": 1,
  "bpe_model": "spm_es.model",
  "bucketing_sampler": true,
  "causal": true,
  "chunk_size": "32",
  "cnn_module_kernel": "31,31,15,15,15,31",
  "concatenate_cuts": false,
  "context_size": 2,
  "ctc_loss_scale": 0.2,
  "decoder_dim": 512,
  "downsampling_factor": "1,2,4,8,4,2",
  "drop_last": true,
  "duration_factor": 1.0,
  "enable_musan": false,
  "enable_spec_aug": true,
  "encoder_dim": "192,256,384,512,384,256",
  "encoder_unmasked_dim": "192,192,256,256,256,192",
  "env_info": {
    "IP address": "",
    "hostname": "",
    "icefall-git-branch": null,
    "icefall-git-date": null,
    "icefall-git-sha1": null,
    "icefall-path": "/zipformer",
    "k2-build-type": "Release",
    "k2-git-date": "Fri Feb 23 01:48:38 2024",
    "k2-git-sha1": "ff1d435a8d3c4eaa15828a84a7240678a70539a7",
    "k2-path": "/envs/zip/lib/python3.10/site-packages/k2/__init__.py",
    "k2-version": "1.24.4",
    "k2-with-cuda": true,
    "lhotse-path": "/envs/zip/lib/python3.10/site-packages/lhotse/__init__.py",
    "lhotse-version": "1.31.1",
    "python-version": "3.10",
    "torch-cuda-available": true,
    "torch-cuda-version": "11.3",
    "torch-version": "1.12.1+cu113"
  },
  "eos_id": 1,
  "exp_dir": "/model/asr-online_10000",
  "feature_dim": 80,
  "feedforward_dim": "512,768,1024,1536,1024,768",
  "gap": 1.0,
  "ignore_id": -1,
  "inf_check": false,
  "input_strategy": "PrecomputedFeatures",
  "joiner_dim": 512,
  "keep_last_k": 30,
  "label_smoothing": 0.1,
  "label_type": "kmeans",
  "left_context_frames": "128",
  "lm_scale": 0.25,
  "log_interval": 50,
  "lr_batches": 7500,
  "lr_epochs": 3.5,
  "manifest_dir": "/zp_datasets/features/es-ar/data/new_fbank_10000",
  "master_port": 12356,
  "max_duration": 500,
  "num_buckets": 30,
  "num_encoder_layers": "2,2,3,4,3,2",
  "num_epochs": 30,
  "num_heads": "4,4,4,8,4,4",
  "num_workers": 2,
  "on_the_fly_feats": false,
  "pos_dim": 48,
  "pos_head_dim": "4",
  "pretrain_path": null,
  "pretrain_type": "ASR",
  "print_diagnostics": false,
  "prune_range": 5,
  "query_head_dim": "32",
  "ref_duration": 600,
  "reset_interval": 200,
  "return_cuts": true,
  "save_every_n": 4000,
  "seed": 1332,
  "shuffle": true,
  "simple_loss_scale": 0.5,
  "sos_id": 1,
  "spec_aug_time_warp_factor": 80,
  "start_batch": 0,
  "start_epoch": 1,
  "subsampling_factor": 4,
  "tensorboard": true,
  "use_attention_decoder": false,
  "use_ctc": false,
  "use_fp16": true,
  "use_transducer": true,
  "valid_interval": 3000,
  "value_head_dim": "12",
  "vocab_size": 4096,
  "warm_step": 2000,
  "world_size": 8
}

**Comment by csukuangfj** [comment]
请问训练收敛了吗，decode.py输出的wer是多少

**Comment by 0120wzw-coder** [comment]
收敛了，wer是14%

**Comment by csukuangfj** [comment]
训练用的特征，和导出onnx模型后，用的特征，是一样的么？如果验证是一样的呢？用训练时提供的特征去跑 onnx 模型，会有问题否？

可以用 pretrained.pt 先去跑一下，我们应该有 pretrained.py 却测试 pretrained.pt ?

**Comment by 0120wzw-coder** [comment]
感谢提供的宝贵意见，上述的问题已经解决，问题的原因是自定义的blank_id与默认的blank_id不一致导致的。
